pkg/watcher/watcher.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watcher
+
+import (
+	"reflect"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/sig-storage-local-static-provisioner/pkg/common"
+)
+
+// ConfigWatcher monitors the config file periodically with the provided interval
+// and compares the provisioner config that is currently applied and the loaded provisioner
+// config. If a difference between the configs is detected, it will signal to the controller
+// to restart.
+type ConfigWatcher struct {
+	configPath        string
+	resyncPeriod      time.Duration
+	lastAppliedConfig common.ProvisionerConfiguration
+}
+
+// NewConfigWatcher creates a new ConfigWatcher object with the provided
+// path of the config file, period to run the load and compare, and the initial
+// configuration that is currently being applied for the provisioner.
+func NewConfigWatcher(configPath string, resyncPeriod time.Duration, config common.ProvisionerConfiguration) *ConfigWatcher {
+	return &ConfigWatcher{
+		configPath:        configPath,
+		resyncPeriod:      resyncPeriod,
+		lastAppliedConfig: config,
+	}
+}
+
+// Run will start running the ConfigWatcher in a loop. During each reload cycle,
+// it loads the configuration from the config file on disk and compares with the last applied
+// configuration. If there is a difference, it will send the loaded configuration to the
+// channel indicating restart controller and then update its last applied configuration.
+func (cw *ConfigWatcher) Run(configUpdate chan<- common.ProvisionerConfiguration) {
+	provisionerConfig := common.ProvisionerConfiguration{
```

Comment:
one nit is that there's duplicated code in main.go to load the config, this file should probably be renamed to watcher/config.go, wdyt?

Reply:
Sure sounds good, changed file name to `watcher/config.go`